### PR TITLE
Trying zend expressive session, csrf

### PR DESCRIPTION
**Pull request by harikt**

### CODE REVIEW

File: src/Auth/HktAuthSystem.php [hunk 1]
```diff
@@ -58,14 +58,14 @@ class HktAuthSystem extends AuthSystem
...
-     * @param Session                $session
+     * @param SessionInterface       $session
      * @param IAdminRepository       $userRepository
      * @param IRoleRepository        $roleRepository
      * @param IPasswordHasherFactory $passwordHasherFactory
      * @param IIocContainer          $iocContainer
      */
     public function __construct(
-        Session $session,
+        SessionInterface $session,
```
Comment (harikt) [inline src/Auth/HktAuthSystem.php:68]: This same SessionInterface is injected to Session middleware

File: src/ContainerConfig.php [hunk 2]
```diff
@@ -85,12 +91,30 @@ public function define(LaravelIocContainer $container)
 
         $container->bindValue(Container::class, $container->getLaravelContainer());
 
-        $container->bindCallback(IIocContainer::SCOPE_SINGLETON, Session::class, function () {
-            $session_factory = new SessionFactory;
-            $session = $session_factory->newInstance($_COOKIE);
-            return $session;
```
Comment (harikt) [inline src/ContainerConfig.php:91]: ServerRequestInterface is immutable, no value is expected / should change. This is not Client Request.

File: src/ContainerConfig.php [hunk 3]
```diff
@@ -85,12 +91,30 @@ public function define(LaravelIocContainer $container)
...
+        $container->bindCallback(IIocContainer::SCOPE_SINGLETON, ServerRequestInterface::class, function () {
+            return ServerRequestFactory::fromGlobals();
         });
 
+        $container->bindCallback(IIocContainer::SCOPE_SINGLETON, SessionPersistenceInterface::class, function () use ($container) {
```
Comment (harikt) [inline src/ContainerConfig.php:98]: SessionPersistenceInterface is injected to Session Middleware

File: src/ContainerConfig.php [hunk 4]
```diff
@@ -85,12 +91,30 @@ public function define(LaravelIocContainer $container)
...
+            return $container->get(PhpSessionPersistence::class);
+        });
+
+        // todo verify alias
+        // $container->alias(PhpSessionPersistence::class, SessionPersistenceInterface::class);
+
+        $container->bindCallback(IIocContainer::SCOPE_SINGLETON, SessionInterface::class, function () use ($container) {
+            return $container->get(LazySession::class);
```
Comment (harikt) [inline src/ContainerConfig.php:106]: LazySession get `SessionPersistenceInterface` and `ServerRequestInterface` .

File: src/Http/Middleware/Session.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        $this->session = $session;
+    }
+
+    public function process(ServerRequestInterface $request, DelegateInterface $delegate) : ResponseInterface
+    {
+        // $session = new LazySession($this->persistence, $request);
+        // var_dump($this->session->get('auth'));
+        // exit;
+        $response = $delegate->process($request);
+        return $this->persistence->persistSession($this->session, $response);
```
Comment (harikt) [inline src/Http/Middleware/Session.php]: Trying to persist the session , added to pipeline.

**Review by harikt — COMMENTED** [review]
@weierophinney this is what I am doing . Don't want to disturb you if you are busy in the middle of conference. You can take your time to respond, but I am interested to hear before you release a final version of zend-expressive-session .

**Comment by harikt** [comment]
Forgot to mention `HktAuthSystem` is injected via constructor to Controllers https://github.com/harikt/web.expressive/tree/master/src/Http/Controllers . All are using constructor injection and I don't like polutting via request with attribute.

File: src/Auth/HktAuthSystem.php (hunk 1 above)
Comment (harikt) [reply src/Auth/HktAuthSystem.php:68]: According to Ocramius https://github.com/zendframework/zend-expressive-session/pull/12 

> Also the session is NOT a service - it is bound to the current request (and specifically some data in the request, usually an identifier).

Now we can't inject a Session ?

This is how zend-expressive-authentication is doing : https://github.com/ezimuel/zend-expressive-authentication/blob/master/src/Adapter/PhpSession.php#L32-L61 via https://github.com/ezimuel/zend-expressive-authentication/blob/42863eae5dcdca07fbfcb1975cd4eb0a9e0a101d/src/AuthenticationMiddleware.php#L37 . The authenticated object is only available at the controller, no way to inject via constructor.

File: src/ContainerConfig.php (hunk 2 above)
Comment (harikt) [reply src/ContainerConfig.php:91]: An internal sub request inside an application is a problem. But does the session, cookie also change on the sub requests ?

**Comment by harikt** [comment]
Closing for this is not feasible or have multiple problems.

**Comment by weierophinney** [comment]
> This is how zend-expressive-authentication is doing 

I'm in the process of extracting the various adapters into their own repositories, and have already identified that adapter as needing refactoring. 

Interestingly, it's not problematic. The session is only needed during the authentication attempt, and the adapter receives the request at that time. As such, I can pull the session from it at that point, and perform well necessary operations.

**Pull request closed without merging**